core/types/calls.cc

```diff
@@ -2919,6 +2937,81 @@ class Magic_mergeHashValues : public IntrinsicMethod {
     }
 } Magic_mergeHashValues;
 
+namespace {
+void convertDefinedToType(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) {
+    // args[0] is the receiver
+    // args[1] is the symbol name of the variable (class or instance)
+    if (args.args.size() != 2) {
+        return;
+    }
+
+    auto &receiver = args.args[0];
+    if (receiver->type.isUntyped()) {
+        res.returnType = receiver->type;
+        return;
+    }
+    if (!receiver->type.isFullyDefined()) {
+        return;
+    }
+
+    auto &classVar = args.args[1];
+    auto classLit = cast_type_nonnull<LiteralType>(classVar->type);
+    auto classVarName = classLit.asName(gs);
+
+    SymbolRef klass;
+    if (isa_type<AppliedType>(receiver->type)) {
+        klass = cast_type_nonnull<AppliedType>(receiver->type).klass;
+        klass = klass.data(gs)->attachedClass(gs);
```

Comment:
Why are these cases not symmetric w.r.t. their treatment of `attachedClass`?

I would have expected to either see them both take attached class on the symbol or to see neither of them.

Do you have test cases for this?

Reply:
Do you mind linking me to those?

Reply:
btw this (well, and the crash posted above) is the thing holding me back from approving this PR. it looks suspicious, so i'm eager to learn more why it's needed.